src/components/legend/draw.js

```diff
@@ -586,23 +586,58 @@ function computeLegendDimensions(gd, groups, traces) {
         extraWidth = 40;
     }
     else if(isGrouped) {
-        var groupXOffsets = [opts._width];
+        var maxHeight = 0;
+        var maxWidth = 0;
         var groupData = groups.data();
 
-        for(var i = 0, n = groupData.length; i < n; i++) {
-            var textWidths = groupData[i].map(function(legendItemArray) {
+        var maxItems = 0;
+
+        groupData.forEach(function(group) {
+
+
+            var groupWidths = group.map(function(legendItemArray) {
                 return legendItemArray[0].width;
             });
 
-            var groupWidth = 40 + Math.max.apply(null, textWidths);
+            var groupWidth = Math.max.apply(null, groupWidths);
+            var groupHeight = group.reduce(function(a, b) {
+                return a + b[0].height;
+            }, 0);
 
-            opts._width += opts.tracegroupgap + groupWidth;
+            maxWidth = Math.max(maxWidth, groupWidth);
+            maxHeight = Math.max(maxHeight, groupHeight);
+            maxItems = Math.max(maxItems, group.length);
+        });
+
+        maxHeight += opts.tracegroupgap || 5;
```

Comment:
When `tracegroupgap` is `0`, which is falsey, it uses value `5` which is not what we want. I would :hocho: `|| 5`.

Reply:
Makes sense, I ripped it from https://github.com/plotly/plotly.js/blob/f279235ed99f3b3fda9805cdfd9b44378003eceb/src/components/legend/draw.js#L672

I guess it was put there to default to 5 which is redundant now it defaults to 10, should I 🔪that also?

Reply:
See https://github.com/plotly/plotly.js/pull/3628#issuecomment-472135605 . Actually, if we reach that block, `opts.tracegroupgap` is still undefined since we only coerce the attribute if the legend has grouped items:
https://github.com/plotly/plotly.js/blob/7e68a2e56f5cfca1286163bda63b5f8263c43451/src/components/legend/defaults.js#L104

Thanks for cleaning that up :smile: !